client/client-multi/client-c/src/kaa_channel_manager.c

```diff
@@ -14,135 +14,283 @@
-        sync_details->sync_fn = NULL;
-        sync_details->supported_services_size = 0;
-        KAA_FREE(sync_details->supported_services);
-        sync_details->supported_services = NULL;
-        KAA_FREE(sync_details);
+    KAA_RETURN_IF_NIL(data,);
+
+    kaa_transport_channel_interface_t *channel =
+            (kaa_transport_channel_interface_t *)data;
+
+    if (channel->release_context) {
+        channel->release_context(channel->context);
     }
+
+    KAA_FREE(channel);
 }
 
-kaa_error_t kaa_channel_manager_create(kaa_channel_manager_t **channel_manager_p, kaa_logger_t *logger)
+
+static bool find_channel(/* current channel */void *data, /* channel-matcher */void *context)
 {
-    KAA_RETURN_IF_NIL(channel_manager_p, KAA_ERR_BADPARAM);
+    KAA_RETURN_IF_NIL2(data, context, false);
+
+    kaa_error_t error_code;
+    kaa_transport_protocol_info_t info1;
+    kaa_transport_protocol_info_t info2;
+
+    kaa_transport_channel_interface_t *channel = (kaa_transport_channel_interface_t *)data;
+    kaa_transport_channel_interface_t *matcher = (kaa_transport_channel_interface_t *)context;
+
+    error_code = channel->get_protocol_info(channel->context, &info1);
+    KAA_RETURN_IF_ERR(error_code);
+    error_code = matcher->get_protocol_info(matcher->context, &info2);
+    KAA_RETURN_IF_ERR(error_code);
+
+    return (0 == memcmp(&info1, &info2, sizeof(kaa_transport_protocol_info_t)));
+}
+
+kaa_error_t kaa_channel_manager_create(kaa_channel_manager_t **channel_manager_p
+                                     , kaa_logger_t *logger)
+{
+    KAA_RETURN_IF_NIL2(channel_manager_p, logger, KAA_ERR_BADPARAM);
 
     *channel_manager_p = (kaa_channel_manager_t *) KAA_MALLOC(sizeof(kaa_channel_manager_t));
     if (!(*channel_manager_p))
         return KAA_ERR_NOMEM;
 
-    (*channel_manager_p)->sync_handlers = NULL;
-    (*channel_manager_p)->on_sync_all = NULL;
-    (*channel_manager_p)->logger = logger;
+    (*channel_manager_p)->transport_channels      = NULL;
+    (*channel_manager_p)->logger                  = logger;
+    (*channel_manager_p)->sync_info.request_id    = 0;
+    (*channel_manager_p)->sync_info.is_up_to_date = false;
+
     return KAA_ERR_NONE;
 }
 
 void kaa_channel_manager_destroy(kaa_channel_manager_t *self)
 {
     if (self) {
-        kaa_list_destroy(self->sync_handlers, destroy_sync_details);
+        kaa_list_destroy(self->transport_channels, destroy_channel);
         KAA_FREE(self);
     }
 }
 
-kaa_error_t kaa_channel_manager_add_sync_handler(kaa_channel_manager_t *self
-        , kaa_sync_handler_fn handler, const kaa_service_t *supported_services, size_t services_count)
+kaa_error_t kaa_channel_manager_add_transport_channel(kaa_channel_manager_t *self
+                                                    , kaa_transport_channel_interface_t *channel)
 {
-    KAA_RETURN_IF_NIL4(self, handler, supported_services, services_count, KAA_ERR_BADPARAM);
+    KAA_RETURN_IF_NIL2(self, channel, KAA_ERR_BADPARAM);
```

Comment:
I believe we should return some channel's ID back to the user. This ID could be calculated using some hash function of callback pointers + context pointer. Also the same ID could be used by the user to remove his channel from the manager. And after all we will use this ID in "find_channel" function.

So the final structure that we are going to store in the list should look like following:
typedef struct {
    uint32_t channel_id;
    kaa_transport_channel_interface_t channel;
} kaa_transport_channel_instance_t;